--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -46,7 +_,7 @@
     private static final ExplosionDamageCalculator EXPLOSION_DAMAGE_CALCULATOR = new ExplosionDamageCalculator();
     private static final int MAX_DROPS_PER_COMBINED_STACK = 16;
     private static final float LARGE_EXPLOSION_RADIUS = 2.0F;
-    private final boolean fire;
+    private boolean fire; // SMPOnline - allow disabling fire
     private final Explosion.BlockInteraction blockInteraction;
     private final ServerLevel level;
     private final Vec3 center;
@@ -475,18 +_,12 @@
         // Paper end - collision optimisations
     }
 
-    private void hurtEntities() {
+    private void hurtEntities(float f, List<Entity> list) {
         if (!(this.radius < 1.0E-5F)) {
-            float f = this.radius * 2.0F;
-            int floor = Mth.floor(this.center.x - f - 1.0);
-            int floor1 = Mth.floor(this.center.x + f + 1.0);
-            int floor2 = Mth.floor(this.center.y - f - 1.0);
-            int floor3 = Mth.floor(this.center.y + f + 1.0);
-            int floor4 = Mth.floor(this.center.z - f - 1.0);
-            int floor5 = Mth.floor(this.center.z + f + 1.0);
-            List <Entity> list = this.level.getEntities(this.excludeSourceFromDamage ? this.source : null, new AABB(floor, floor2, floor4, floor1, floor3, floor5), entity -> entity.isAlive() && !entity.isSpectator()); // Paper - Fix lag from explosions processing dead entities, Allow explosions to damage source
+            // SMPOnline - move prep before method
+
             for (Entity entity : list) { // Paper - used in loop
-                if (!entity.ignoreExplosion(this)) {
+                if (true) { // !entity.ignoreExplosion(this)) { // SMPOnline - handle elsewhere
                     double d = Math.sqrt(entity.distanceToSqr(this.center)) / f;
                     if (!(d > 1.0)) {
                         Vec3 vec3 = entity instanceof PrimedTnt ? entity.position() : entity.getEyePosition();
@@ -516,6 +_,27 @@
                                         dragonPart.hurtServer(this.level, this.damageSource, this.damageCalculator.getEntityDamageAmount(this, dragonPart, f1));
                                     }
                                 }
+                                // Canvas start - chain end crystal explosions config
+                            } else if (io.canvasmc.canvas.Config.INSTANCE.chainEndCrystalExplosions && entity instanceof net.minecraft.world.entity.boss.enderdragon.EndCrystal endCrystal) {
+                                int chunkX = endCrystal.chunkPosition().x;
+                                int chunkZ = endCrystal.chunkPosition().z;
+                                io.papermc.paper.threadedregions.RegionizedServer.getInstance().taskQueue.queueTickTaskQueue(
+                                    (ServerLevel) endCrystal.level(),
+                                    chunkX, chunkZ, () -> {
+                                        // copied from EndCrystal#hurtServer
+                                        // we don't modify the method directly in EndCrystal#hurtServer, because the chain would execute *instantly* which we do *NOT* want
+                                        // as this can cause major lag spikes given how large the chain is, so spreading this out over multiple ticks is a more efficient idea
+                                        DamageSource damageSource1 = damageSource.getEntity() != null ? endCrystal.damageSources().explosion(endCrystal, damageSource.getEntity()) : null;
+                                        org.bukkit.event.entity.ExplosionPrimeEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.callExplosionPrimeEvent(endCrystal, 6.0F, false);
+                                        if (event.isCancelled()) {
+                                            return;
+                                        }
+
+                                        endCrystal.remove(Entity.RemovalReason.KILLED, org.bukkit.event.entity.EntityRemoveEvent.Cause.EXPLODE); // Paper - add Bukkit remove cause
+                                        level.explode(endCrystal, damageSource1, null, endCrystal.getX(), endCrystal.getY(), endCrystal.getZ(), event.getRadius(), event.getFire(), Level.ExplosionInteraction.BLOCK);
+                                    }
+                                );
+                                // Canvas end - chain end crystal explosions config
                             } else {
                                 entity.hurtServer(this.level, this.damageSource, this.damageCalculator.getEntityDamageAmount(this, entity, f1));
                             }
@@ -529,7 +_,7 @@
                         double d1 = entity instanceof LivingEntity livingEntity
                             ? livingEntity.getAttributeValue(Attributes.EXPLOSION_KNOCKBACK_RESISTANCE)
                             : 0.0;
-                        double d2 = entity instanceof Player && this.level.paperConfig().environment.disableExplosionKnockback ? 0 : (1.0 - d) * f1 * knockbackMultiplier * (1.0 - d1); // Paper
+                        double d2 = entity instanceof Player && this.level.paperConfig().environment.disableExplosionKnockback ? 0 : canvas$getModifiableExplosionKnockback(entity, (1.0 - d) * f1 * knockbackMultiplier, d1); // Paper // Canvas - old blast protection
                         Vec3 vec32 = vec31.scale(d2);
                         // CraftBukkit start - Call EntityKnockbackEvent
                         if (entity instanceof LivingEntity) {
@@ -552,20 +_,49 @@
             }
         }
     }
-
-    private void interactWithBlocks(List<BlockPos> blocks) {
-        List<ServerExplosion.StackCollector> list = new ArrayList<>();
-        Util.shuffle(blocks, this.level.random);
-
-        // CraftBukkit start
-        Location location = CraftLocation.toBukkit(this.center, this.level);
-        List<org.bukkit.block.Block> blockList = new ObjectArrayList<>();
-        for (int i1 = blocks.size() - 1; i1 >= 0; i1--) {
-            org.bukkit.block.Block bblock = org.bukkit.craftbukkit.block.CraftBlock.at(this.level, blocks.get(i1));
-            if (!bblock.getType().isAir()) {
-                blockList.add(bblock);
+    // Canvas start - old blast protection logic
+
+    private static final net.minecraft.world.entity.EquipmentSlot[] SLOTS = {
+        net.minecraft.world.entity.EquipmentSlot.HEAD,
+        net.minecraft.world.entity.EquipmentSlot.CHEST,
+        net.minecraft.world.entity.EquipmentSlot.LEGS,
+        net.minecraft.world.entity.EquipmentSlot.FEET
+    };
+
+    public static double canvas$getModifiableExplosionKnockback(Entity entity, double v, double resistance) {
+        if (!io.canvasmc.canvas.Config.INSTANCE.combat.useOldBlastProtection) {
+            return v * (1.0 - resistance);
+        }
+        // if not living entity, just return velocity because only living
+        // entities are affected by explosion knockback resistance
+        if (!(entity instanceof LivingEntity livingEntity)) {
+            return v;
+        }
+        // 1.20.6 logic
+        int highest = 0;
+
+        for (net.minecraft.world.entity.EquipmentSlot slot : SLOTS) {
+            ItemStack itemStack = livingEntity.getItemBySlot(slot);
+            if (!itemStack.isEmpty()) {
+                int level = net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(
+                    net.minecraft.server.MinecraftServer.getServer().registryAccess().getOrThrow(net.minecraft.world.item.enchantment.Enchantments.BLAST_PROTECTION), itemStack
+                );
+                if (level > highest) {
+                    highest = level;
+                }
             }
         }
+
+        if (highest > 0) {
+            v *= Mth.clamp(1.0 - (double) highest * 0.15, 0.0F, 1.0F);
+        }
+
+        return v;
+    }
+    // Canvas end - old blast protection logic
+
+    private void interactWithBlocks(List<ServerExplosion.StackCollector> list, Location location, List<BlockPos> blocks, List<org.bukkit.block.Block> blockList) {
+
 
         List<org.bukkit.block.Block> bukkitBlocks;
 
@@ -639,14 +_,72 @@
         this.mutablePos = new BlockPos.MutableBlockPos();
         // Paper end - collision optimisations
         this.level.gameEvent(this.source, GameEvent.EXPLODE, this.center);
-        List<BlockPos> list = this.calculateExplodedPositions();
-        this.hurtEntities();
-        if (this.interactsWithBlocks()) {
-            this.interactWithBlocks(list);
+
+        List<BlockPos> blocks = this.calculateExplodedPositions();
+        float f = this.radius * 2.0F;
+        int floor = Mth.floor(this.center.x - f - 1.0);
+        int floor1 = Mth.floor(this.center.x + f + 1.0);
+        int floor2 = Mth.floor(this.center.y - f - 1.0);
+        int floor3 = Mth.floor(this.center.y + f + 1.0);
+        int floor4 = Mth.floor(this.center.z - f - 1.0);
+        int floor5 = Mth.floor(this.center.z + f + 1.0);
+        List <Entity> entities = this.level.getEntities(
+            this.excludeSourceFromDamage ? this.source : null,
+            new AABB(floor, floor2, floor4, floor1, floor3, floor5),
+            entity -> entity.isAlive() && !entity.isSpectator() && !entity.ignoreExplosion(this)
+        ); // Paper - Fix lag from explosions processing dead entities, Allow explosions to damage source
+
+        List<org.bukkit.block.Block> blockList = null;
+        List<ServerExplosion.StackCollector> list = new java.util.ArrayList<>();
+        Location location = CraftLocation.toBukkit(this.center, this.level);
+        org.bukkit.block.Block block = null;
+        org.bukkit.block.BlockState blockState = null;
+        if(this.interactsWithBlocks()){
+            Util.shuffle(blocks, this.level.random);
+
+            // CraftBukkit start
+            blockList = new ObjectArrayList<>();
+            for (int i1 = blocks.size() - 1; i1 >= 0; i1--) {
+                org.bukkit.block.Block bblock = org.bukkit.craftbukkit.block.CraftBlock.at(this.level, blocks.get(i1));
+                if (!bblock.getType().isAir()) {
+                    blockList.add(bblock);
+                }
+            }
+
+            block = location.getBlock();
+            blockState = (this.damageSource.causingBlockSnapshot() != null) ? this.damageSource.causingBlockSnapshot() : block.getState();
         }
 
-        if (this.fire) {
-            this.createFire(list);
+        net.smponline.event.PreExplodeEvent preExplodeEvent = CraftEventFactory.callPreExplodeEvent(
+            this.source == null ? null : this.source.getBukkitEntity(),
+            block, blockState,
+            blockList,
+            new java.util.ArrayList<>(entities.stream().map(e -> (org.bukkit.entity.Entity)e.getBukkitEntity()).toList()),
+            this.yield,
+            this.fire,
+            this.getBlockInteraction()
+        );
+
+        if(!preExplodeEvent.isCancelled()) {
+            this.fire = preExplodeEvent.isCauseFire();
+            entities = new java.util.ArrayList<>(preExplodeEvent.getAffectedEntities().stream().map(e -> ((org.bukkit.craftbukkit.entity.CraftEntity)e).getHandle()).toList());
+
+            // reset blocks to sane list
+            blocks.clear();
+            blockList = preExplodeEvent.getAffectedBlocks();
+            for (org.bukkit.block.Block bblock : blockList) {
+                blocks.add(((org.bukkit.craftbukkit.block.CraftBlock) bblock).getPosition());
+            }
+
+
+            this.hurtEntities(f, entities);
+            if (this.interactsWithBlocks()) {
+                this.interactWithBlocks(list, location, blocks, blockList);
+            }
+
+            if (this.fire) {
+                this.createFire(blocks);
+            }
         }
 
         // Paper start - collision optimisations
@@ -656,7 +_,7 @@
         this.directMappedBlockCache = null;
         this.mutablePos = null;
         // Paper end - collision optimisations
-        return list.size();
+        return blocks.size();
     }
 
     private static void addOrAppendStack(List<ServerExplosion.StackCollector> stackCollectors, ItemStack stack, BlockPos pos) {
